### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by attilaeross**

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -131,22 +143,14 @@ export default function createApp(rootElement) {
...
-      todoItems = getStoredTodos(userName);
-
-      removeAllTodoElements();
-
-      renderAll(todoItems);
+  const setUserTodos = () => {
+    todoItems = getStoredTodos();
+    removeAllTodoElements();
+    renderAll(todoItems);
+  };
```
Comment (mhipszki) [inline src/app.js:152]: would this be better called as `restoreTodos`?

File: src/app.test.js [hunk 2]
```diff
@@ -40,16 +41,36 @@ const addTodo = (todoList, text, completed) => {
...
 
-test("renders empty todo list when starts", () => {
+const logInUser = (userName) => {
+  const userInput = screen.getByPlaceholderText("Please enter username");
+  const setUserButton = screen.getByText("Set User");
+
+  userEvent.type(userInput, userName);
+  userEvent.click(setUserButton);
+};
+test("wont allow to add new todo until user is not set", () => {
```
Comment (mhipszki) [inline src/app.test.js]: an empty line can improve readability

Comment (mhipszki) [reply src/app.test.js]: test case names are best when simplest e.g.:

```js
test("does not allow adding new todo until user is not set", () => {
```

File: src/app.test.js [hunk 3]
```diff
@@ -40,16 +41,36 @@ const addTodo = (todoList, text, completed) => {
...
+  // act
+  const todoList = screen.getByTestId("todo-list");
+  addTodo(todoList, "Take wife for a walk!");
+
+  // assert
+  expect(
+    queryByDisplayValue(todoList, "Take wife for a walk!")
+  ).not.toBeInTheDocument();
+});
```
Comment (mhipszki) [inline src/app.test.js:56]: this is interesting, I think checking that `addTodo` does not result in a particular todo added allows too much room and can create a false positive scenario

I believe the code could be change in a way that this test would still pass but a todo item would still be added to the list e.g. the code amends the user input and prefixes it with a `*` or something similar

in this case it'd be better to check the concrete input and button elements that they are disabled which means it is _not possible_ to use them

File: src/app.test.js [hunk 4]
```diff
@@ -40,16 +41,36 @@ const addTodo = (todoList, text, completed) => {
...
+
+test("renders empty todo list when new user is logs in", () => {
   // setup
+  logInUser("Attila");
```
Comment (mhipszki) [inline src/app.test.js]: this has become a duplicate now:

```js
logInUser("Attila");
```

as it's been repeated in almost every test case

let's use the parameter default value in `logInUser` instead and only pass in the username when the scenario requires a different one

Comment (mhipszki) [reply src/app.test.js]: btw the "login" step is actually an "act" rather then part of the "setup" phase

File: src/app.test.js [hunk 5]
```diff
@@ -59,8 +80,9 @@ test("shows new item in todo list when added", () => {
   ).toBeInTheDocument();
 });
 
-test("removes an item from todo List when user clicks Delete button", () => {
+test("removes an item from todo List when logged in user clicks Delete button", () => {
```
Comment (mhipszki) [inline src/app.test.js]: I don't think we need to add "logged in" to any of the test case names as it is obvious that none of the Todo App functionality can be used without "logging in" first

File: src/app.test.js (hunk 4 above)
Comment (mhipszki) [reply src/app.test.js]: once it's done, I'd have a think about if we could move this to `beforeEach` as almost all test cases need it - apart from the first one which is checking that without logging in the user cannot add a todo

the question is how can we let the application "log out" the current user? :)

File: src/app.test.js [hunk 6]
```diff
@@ -180,16 +201,31 @@ test("shows outstanding todos when filter is set to Outstanding", () => {
...
- * add few todos
- * destroy todo app
- * ACT
- * re render todo app
- * input same user name
- * ASSERT
- * see previously added todos rendered
- */
-test.todo("renders todo list for existing user");
+test("renders todo list for existing user", () => {
```
Comment (mhipszki) [inline src/app.test.js]: is this really what it is? :)

what I see here:

- user "A" logs in, adds 2 todos
- (a new) user "B" logs in (todo list should be emptied)
- user "A" logs back in and the belonging todos are loaded again

btw what does "local assert" mean here?

File: src/app.test.js [hunk 7]
```diff
@@ -40,16 +41,37 @@ const addTodo = (todoList, text, completed) => {
...
+  const todoList = screen.getByTestId("todo-list");
+  addTodo(todoList, "Take wife for a walk!");
+
+  // assert
+  expect(
+    queryByDisplayValue(todoList, "Take wife for a walk!")
+  ).not.toBeInTheDocument();
+});
+
+test("renders empty todo list when new user is logs in", () => {
```
Comment (mhipszki) [inline src/app.test.js]: typo

File: src/app.test.js [hunk 8]
```diff
@@ -180,16 +202,32 @@ test("shows outstanding todos when filter is set to Outstanding", () => {
...
- * add few todos
- * destroy todo app
- * ACT
- * re render todo app
- * input same user name
- * ASSERT
- * see previously added todos rendered
- */
-test.todo("renders todo list for existing user");
+test("renders empty todo list when changing to a nonexistent user", () => {
```
Comment (mhipszki) [inline src/app.test.js]: I believe there is no such a thing as "user" atm, or maybe there is but just on a theoretical level

what we've got is a bunch of todo items stored in `localStorage` where the key is generated from the "username", maybe "nonexistent user" would be better to use e.g. as "new user" here in the wording?

File: src/app.test.js [hunk 9]
```diff
@@ -180,16 +202,32 @@ test("shows outstanding todos when filter is set to Outstanding", () => {
...
+test("renders todo list for existing user", () => {
+  // setup
+  logInUser();
+  const todoList = screen.getByTestId("todo-list");
+  addTodo(todoList, "Take wife for a walk!", true);
+  addTodo(todoList, "Take dog for a walk!");
+
+  // act
+  logInUser("Marton");
+  logInUser();
```
Comment (mhipszki) [inline src/app.test.js]: this is the case where it would be better to _explicitly_ give the username to `logInUser`, otherwise the reader can get confused and forced to go and check what's happening in `logInUser` just to find out there's a default username "Attila"

for demonstration we could just use "A" and "B" to improve comprehensiveness

also, it can be quite difficult to distinguish between "Take wife for a walk!" and "Take dog for a walk!" as they differ only in a few characters, would be much better to use e.g. "Todo 1 of user A", "Todo 2 of user A" etc. and "Todo 1 of user B", "Todo 2 of user B" respectively

Comment (mhipszki) [reply src/app.test.js]: actually, thinking of this "log in", it is essentially a "change/switch user" action rather than a "log in" action, give it a try and rename it to see if that reads better in the test cases e.g. `changeUserTo("A")` or `switchUserTo("B")`

File: src/app.test.js [hunk 10]
```diff
@@ -40,16 +41,36 @@ const addTodo = (todoList, text, completed) => {
...
+  const setUserButton = screen.getByText("Set User");
+
+  userEvent.type(userInput, userName);
+  userEvent.click(setUserButton);
+};
+
+test("does not allow to enter / add new todo until user is not set", () => {
+  // act
   const todoList = screen.getByTestId("todo-list");
+  addTodo(todoList, "Take wife for a walk!");
```
Comment (mhipszki) [inline src/app.test.js]: how does "adding" a todo item help when input / button is disabled?

File: src/app.test.js [hunk 11]
```diff
@@ -40,16 +41,36 @@ const addTodo = (todoList, text, completed) => {
...
+  // setup
   // act
+  setUser();
+  const todoList = screen.getByTestId("todo-list");
   // assert
   expect(todoList).toBeEmptyDOMElement();
 });
 
-test("shows new item in todo list when added", () => {
+test("shows new item in todo list for when added", () => {
```
Comment (mhipszki) [inline src/app.test.js]: typo

File: src/app.test.js [hunk 12]
```diff
@@ -180,16 +201,32 @@ test("shows outstanding todos when filter is set to Outstanding", () => {
...
+
+test("renders todo list for existing user", () => {
+  // setup
+  setUser("A");
+  const todoList = screen.getByTestId("todo-list");
+  addTodo(todoList, "Todo 1 for user A", true);
+  addTodo(todoList, "Todo 2 for user A");
+
+  // act
+  setUser("B");
```
Comment (mhipszki) [inline src/app.test.js:222]: as a safety check I'd add the assertion here as well:

`expect(todoList).toBeEmptyDOMElement();`

then set user "A"

File: src/app.test.js (hunk 10 above)
Comment (attilaeross) [reply src/app.test.js]: Wont help in this instance as we don't check if this will appear in the list....was used in the previous assertion
